pennylane/compiler.py

```diff
@@ -0,0 +1,227 @@
+# Copyright 2023 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""This subpackage provides the JIT compilation support for PennyLane.
+
+
+This subpackage exclusively functions as a wrapper for PennyLane's JIT compiler packages, without independently
+implementing any compiler itself. Currently, it supports the ``pennylane-catalyst`` package, with plans to
+incorporate additional packages in the near future.
+
+For any compiler packages seeking to be registered, it is imperative that they expose the 'entry_points' metadata
+under the designated group name: ``pennylane.compilers``.
+"""
+
+from typing import List
+from importlib import reload
+import pkg_resources
+
+
+class Compiler:
+    """The JIT compiler module containing all utilities and wrapper methods for ``qml.qjit`` in PennyLane."""
+
+    # Private class properties
+    __BACKENDS = []
+    __ENTRY_POINTS = {}
+
+    @classmethod
+    def available(cls, name: str = "pennylane-catalyst") -> bool:
+        """Check the availability of the given compiler package.
+
+        Args:
+            name (str): the name of the compiler package
+
+        Return:
+            bool : True if the compiler package is installed on the system
+        """
+
+        reload(pkg_resources)
+        pkg_name = pkg_resources.Requirement.parse(name)
+
+        if pkg_resources.working_set.find(pkg_name):
+            if name not in cls.__BACKENDS:
+                cls.__ENTRY_POINTS = {
+                    entry.name: entry
+                    for entry in pkg_resources.iter_entry_points("pennylane.compilers")
+                }
+                cls.__BACKENDS.append(name)
+            return True
+
+        if name in cls.__BACKENDS:
+            cls.__BACKENDS.remove(name)
```

Comment:
This branch is curious, does it cover cases where the compiler package has been found before but was then unloaded?

Reply:
Yes, this captures corner cases that may trigger by installing and uninstalling Catalyst in a notebook. Also check my answer for `reload(pkg_resources)`.

Reply:
@maliasadi is is the same edge case that we're tackling with devices here?

https://github.com/PennyLaneAI/pennylane/blob/21ec4d4ed30bde24894f3db5a9a38222ea05db1d/pennylane/__init__.py#L147-L159